src/Analyzers/Core/Analyzers/NamespaceSync/AbstractNamespaceSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+        /// </summary>
+        /// <returns></returns>
+        protected abstract bool ContainsPartialTypeWithMultipleDeclarations(TNamespaceSyntax namespaceDeclaration, SemanticModel semanticModel);
+
+        /// <summary>
+        /// Gets the syntax representing the name of a namespace declaration
+        /// </summary>
+        protected abstract SyntaxNode GetNameSyntax(TNamespaceSyntax namespaceDeclaration);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            // It's ok to not have a rootnamespace property, but if it's there we want to use it correctly
+            context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace);
+
+            // Project directory is a must to correctly get the relative path and construct a namespace
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || string.IsNullOrEmpty(projectDir))
+            {
+                return;
+            }
+
+            if (context.Node is TNamespaceSyntax namespaceDecl)
+            {
+                var currentNamespace = GetNamespaceName(namespaceDecl, rootNamespace);
+
+                if (IsFileAndNamespaceMismatch(namespaceDecl, rootNamespace, projectDir, out var targetNamespace) &&
+                    IsFixSupported(context.SemanticModel, namespaceDecl, context.CancellationToken))
+                {
+                    var nameSyntax = GetNameSyntax(namespaceDecl);
+                    AbstractNamespaceSyncDiagnosticAnalyzer<TNamespaceSyntax>.ReportDiagnostics(context, Descriptor, nameSyntax.GetLocation(), targetNamespace, currentNamespace);
+                }
+            }
+        }
+
+        private bool IsFixSupported(SemanticModel semanticModel, TNamespaceSyntax namespaceDeclaration, CancellationToken cancellationToken)
+        {
+            var root = namespaceDeclaration.SyntaxTree.GetRoot(cancellationToken);
+
+            // It should not be nested in other namespaces
+            if (namespaceDeclaration
+                .Ancestors()
+                .OfType<TNamespaceSyntax>()
+                .Any())
+            {
+                return false;
+            }
+
+            // It should not contain a namespace
+            var containsNamespace = namespaceDeclaration
+                 .DescendantNodes(n => n is TNamespaceSyntax)
+                 .OfType<TNamespaceSyntax>().Any();
+            if (containsNamespace)
+            {
+                return false;
+            }
+
+            // It should not contain partial classes with more than one instance in the semantic model. The
+            // fixer does not support this scenario.
```

Comment:
i'm curious why not.  why would thismatter?  we're not changing the namespace right?  we're changing the folder structure.

(or did i get that reversed?  if i did, we need to update these type names to be clear as to what they do).

Reply:
You got it reversed. We're changing the namespace not the folder structure. I filed https://github.com/dotnet/roslyn/issues/50686 to cover the other fix scenario

Reply:
it's hard to tell with the name 'NamespaceSync'.  If it was 'UpdateNamespaceToMatchFolderStructure' it would be clearer :)

Reply:
Changed naming of everything. Let me know what you think